vnet/network_conditioner.go

```diff
@@ -0,0 +1,132 @@
+package vnet
+
+import (
+	"math/rand"
+	"sync/atomic"
+	"time"
+	"unsafe"
+)
+
+const (
+	// NetworkConditionPacketLossDenominator is the denominator by which the `NetworkCondition.PacketLoss` value
+	// will be divided.
+	NetworkConditionPacketLossDenominator uint32 = 1000
+
+	kiloBytesInBytes = 1000
+	// Value chosen arbitrarily, there may be a better one.
+	bandwidthComputeThreshold = 2 * time.Second
+)
+
+// NetworkConditioner hold the NetworkCondition of DownLink and UpLink channels.
+type NetworkConditioner struct {
+	// Internals stats
+
+	// Total of downLink bytes
+	// Must be accessed atomically.
+	inBytes                  uint64
+	pendingInBytes           uint64
+	lastInByteChunkTimestamp int64
+
+	// Total of upLink bytes
+	// Must be accessed atomically.
+	outBytes                  uint64
+	pendingOutBytes           uint64
+	lastOutByteChunkTimestamp int64
+
+	// DownLink represents the DownLink condition.
+	DownLink NetworkCondition
+	// UpLink represents the UpLink condition.
+	UpLink NetworkCondition
+}
+
+func (n *NetworkConditioner) handleDownLink(c Chunk) bool {
+	atomic.AddUint64(&n.pendingInBytes, uint64(len(c.UserData())))
+
+	bandwidthLimiterSleep := n.bandwidthLimiter(&n.DownLink.MaxBandwidth, &n.lastInByteChunkTimestamp, &n.inBytes, &n.pendingInBytes)
+
+	// Apply constant latency if required
+	if sleepNeeded := time.Duration(atomic.LoadInt64((*int64)(&n.DownLink.Latency))) - bandwidthLimiterSleep; sleepNeeded > 0 {
+		time.Sleep(sleepNeeded)
```

Comment:
If I'm not mistaken (it's been a while since I wrote this code.. :P), you are blocking the calling thread which is the `processChunks` routine from a router. During this blockage, the route cannot process other chunks unrelated to this NIC.  I believe this is not what you intended...